govcd/api.go

```diff
@@ -178,3 +178,100 @@ func checkResp(resp *http.Response, err error) (*http.Response, error) {
 		return nil, fmt.Errorf("unhandled API response, please report this issue, status code: %s", resp.Status)
 	}
 }
+
+// Helper function creates request, run it, check response and parse task from response.
+// pathUlr - request URL
+// requestType - HTTP method type
+// contentType - value which will be set for "Content-Type"
+// errorMessage - error message which to return when error happen
+// payload - XML struct which will be marshalled and added as body/payload
+// E.g. client.ExecuteTaskRequest(updateDiskLink.HREF, http.MethodPut, updateDiskLink.Type, "error update disk: %s", xmlPayload)
+func (client *Client) ExecuteTaskRequest(pathURL, requestType, contentType, errorMessage string, payload interface{}) (Task, error) {
+
+	resp, err := executeRequest(pathURL, requestType, contentType, errorMessage, payload, client)
+	if err != nil {
+		return Task{}, fmt.Errorf(errorMessage, err)
+	}
+
+	task := NewTask(client)
+
+	if err = decodeBody(resp, task.Task); err != nil {
+		return Task{}, fmt.Errorf("error decoding Task response: %s", err)
+	}
+
+	resp.Body.Close()
+
+	// The request was successful
+	return *task, nil
+}
+
+// Helper function creates request, run it, check response and do not expect any values from it.
+// pathUlr - request URL
+// requestType - HTTP method type
+// contentType - value which will be set for "Content-Type"
+// errorMessage - error message which to return when error happen
+// payload - XML struct which will be marshalled and added as body/payload
+// E.g. client.ExecuteRequestWithoutResponse(catalogItemHREF.String(), http.MethodDelete, "", "error deleting Catalog item: %s", nil)
+func (client *Client) ExecuteRequestWithoutResponse(pathURL, requestType, contentType, errorMessage string, payload interface{}) error {
+
+	resp, err := executeRequest(pathURL, requestType, contentType, errorMessage, payload, client)
+	if err != nil {
+		return fmt.Errorf(errorMessage, err)
+	}
+
+	resp.Body.Close()
+
+	// The request was successful
+	return nil
+}
+
+// Helper function creates request, run it, check response and parse task from response.
+// pathUlr - request URL
+// requestType - HTTP method type
+// contentType - value which will be set for "Content-Type"
+// errorMessage - error message which to return when error happen
+// payload - XML struct which will be mashalled and added as body/payload
+// out - structure to be used for unmarshalling xml
+// E.g. 	unmarshalledAdminOrg := &types.AdminOrg{}
+// client.ExecuteRequest(adminOrg.AdminOrg.HREF, http.MethodGet, "", "error refreshing organization: %s", nil, unmarshalledAdminOrg)
+func (client *Client) ExecuteRequest(pathURL, requestType, contentType, errorMessage string, payload, out interface{}) error {
```

Comment:
Same considerations as given for `ExecuteTaskRequest`. We need to return a response.

Reply:
Response is marshalled to out interface.